File: Medium/114. Flatten Binary Tree to Linked List/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def flatten(self, root):
        """
        :type root: TreeNode
        :rtype: void Do not return anything, modify root in-place instead.
        """
        pre_v = root
        for v in self.preorder(root):
            if v is pre_v:
                continue
            pre_v.left = None
            pre_v.right = v
            pre_v = v
    
    def preorder(self, root):
        if not root:
            return None
        left, right = root.left, root.right
        yield root
        yield from self.preorder(left)
        yield from self.preorder(right)

File: Medium/114. Flatten Binary Tree to Linked List/solution.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {void} Do not return anything, modify root in-place instead.
 */
var flatten = function(root) {
    let pre_value = root
    for (const value of preorder(root)) {
        if (value === pre_value)
            continue
        pre_value.left = null
        pre_value.right = value
        pre_value = value
    }
};

function *preorder(root) {
    if (!root)
        return
    const {left, right} = root
    yield root
    yield* preorder(left)
    yield* preorder(right)
}